trace/obfuscate.go

```diff
@@ -22,19 +55,148 @@ type obfuscationVisitor struct {
+
+// EnterData processes the given data and obfuscates sensitive information based on the provided obfuscation options.
+// It does 3 checks to determine if the data contains sensitive information:
+// 1. It checks if spec is an HTTP Authorization or Cookie data and obfuscates the value.
+// 2. It checks if spec is an HTTP Header or Query Param data and obfuscates the value if the key is in the list of sensitive keys.
+// 3. It applies regex patterns to all obfuscate sensitive primitive string values.
+func (s *redactSensitiveInfoVisitor) EnterData(self interface{}, _ vis.SpecVisitorContext, d *pb.Data) Cont {
+	if httpMeta := d.GetMeta().GetHttp(); httpMeta != nil {
+		var key string
+		switch httpMeta.Location.(type) {
+		case *pb.HTTPMeta_Auth:
+			return ObfuscatePrimitiveWithRedactedString(d)
+		case *pb.HTTPMeta_Cookie:
+			return ObfuscatePrimitiveWithRedactedString(d)
+		case *pb.HTTPMeta_Header:
+			header := httpMeta.GetHeader()
+			key = header.Key
+		case *pb.HTTPMeta_Query:
+			queryParam := httpMeta.GetQuery()
+			key = queryParam.Key
+		}
+		// Check if the key is in the list of keys to obfuscate.
+		if s.obfuscationOptions.SensitiveDataKeys.Contains(strings.ToLower(key)) {
+			return ObfuscatePrimitiveWithRedactedString(d)
+		}
+	}
+
+	primitive := d.GetPrimitive()
+	if primitive == nil {
+		// Not a primitive, regex will be applied to primitive string values only.
+		return Continue
+	}
+
+	stringValue := primitive.GetStringValue()
+	if stringValue == nil {
+		// Not a string, regex will be applied to string values only.
+		return Continue
+	}
+
+	for _, pattern := range s.obfuscationOptions.SensitiveDataValuePatterns {
+		if pattern.MatchString(stringValue.Value) {
+			return ObfuscatePrimitiveWithRedactedString(d)
+		}
+	}
+
+	return Continue
+}
+
+func (s *redactSensitiveInfoVisitor) EnterHTTPBody(self interface{}, ctx vis.SpecVisitorContext, b *pb.HTTPBody) Cont {
+	node, _ := ctx.GetInnermostNode(reflect.TypeOf((*pb.Data)(nil)))
+	data := node.(*pb.Data)
+	if data == nil {
+		return Continue
+	}
+
+	// Traverse the data node and redact sensitive information.
+	s.traverseAndRedactSensitiveInfo(data, false)
+
+	return Continue
+}
+
+// Traverses the protobuf data and redacts sensitive information based on the sensitive keys.
+// The function has 2 parameters:
+//  1. data: The current node in the recursive tree traversal of the protobuf data.
+//  2. redactData: A boolean flag to indicate if the data should be redacted irrespective of it's key and value.
+//     This will be true when the parent node is a sensitive key.
+//
+// The function checks if the data fields are primitive and redacts them if their keys are sensitive.
+// For non-primitive fields, it recursively traverses it's child nodes with the redactData flag set to true if the parent node is sensitive.
+func (s *redactSensitiveInfoVisitor) traverseAndRedactSensitiveInfo(data *pb.Data, redactData bool) {
```

Comment:
I had hoped to have a unit test of this specifically so that we could test the error cases, which we otherwise believe are unreachable.